[Workers] Jamstack tutorial revision (#2131)
cloudflare/cloudflare-docs: products/workers/src/content/tutorials/build-a-jamstack-app/index.md

import TutorialsBeforeYouStart from "../../_partials/_tutorials-before-you-start.md"

> # Build a To-Do list JAMstack application

Author: ```suggestion
# Build a Todo list JAMstack application
```

$ cd todos
```

Wrangler templates are just Git repositories, so if you want to create your own templates, or use one from the [Template Gallery](/examples), there are various options to help you get started.

> Wrangler’s default template includes support for building and deploying JavaScript-based projects, including Webpack support. Inside of your new `todos` directory, `index.js` represents the entry-point to your Cloudflare Workers application.

Author: ```suggestion
Wrangler’s default template includes support for building and deploying JavaScript-based projects, including Webpack support. Inside of your new `todos` directory, `index.js` represents the entry point to your Cloudflare Workers application.
```

}
```

The Worker KV data store is eventually consistent: writing to the cache means that it will become available eventually, but it is possible to attempt to read a value back from the cache immediately after writing it, only to find that the cache has not been updated yet.

> Given the presence of data in the cache, and the assumption that your cache is eventually consistent, you should adjust this code slightly: first, you should actually read from the cache, parsing the value back out, and using it as if the data source exists. If it does not, you will refer to `defaultData`, setting it as the data source for now (remember, it should be set in the future), while also setting it in the cache for future use. After breaking out the code into a few functions for simplicity, the result looks like this:

Reviewer: ```suggestion
Given the presence of data in the cache and the assumption that your cache is eventually consistent, this code needs a slight adjustment: the application should check the cache and use its value, if the key existed. If it does not, you will use `defaultData` as the data source for now (remember, it should be set in the future) and write it to the cache for future use. After breaking out the code into a few functions for simplicity, the result looks like this:
```

### Rendering data from KV

Given the presence of data in your code, which is the cached data object for your application, take this data and make it available on screen.

> In your Workers script, make a new variable, `html`, and use it to build up a static HTML template that you can serve to the client. In `handleRequest`, construct a new `Response` (with a `Content-Type` header of `text/html`), and serve it to the client:

Reviewer: ```suggestion
In your Workers script, make a new `html` variable and use it to build up a static HTML template that you can serve to the client. In `handleRequest`, construct a new `Response` with a `Content-Type: text/html` header and serve it to the client:
```

}
```

> You have a static HTML site being rendered, and now you can begin populating it with data. In the body, add a `div` tag with an id of `todos`:

Reviewer: ```suggestion
You have a static HTML site being rendered and now you can begin populating it with data. In the body, add a `div` tag with an id of `todos`:
```

`
```

> Add a script after the body that takes a `todos` array, and loops through it, and for each todo in the array, creates a `div` element and appends it to the list of todos:

Reviewer: ```suggestion
Add a `<script>` after the body that takes a `todos` array. For each `todo` in the array, create a `div` element and appends it to the `todos` HTML element:
```

`
```

> Add a `<script>` after the body that takes a `todos` array. For each `todo` in the array, create a `div` element and appends it to the `todos` HTML element:

Reviewer: How about `at the end of the end of the body content`?

> Add a `<script>` after the body that takes a `todos` array. For each `todo` in the array, create a `div` element and appends it to the `todos` HTML element:

Reviewer 2: Updated my suggestion to `at the end of the body content`.

`
```

> Your static page can take in `window.todos`, and render HTML based on it, but you have not actually passed in any data from KV. To do this, you will need to make a couple changes.

Reviewer: ```suggestion
Your static page can take in `window.todos` and render HTML based on it, but you have not actually passed in any data from KV. To do this, you will need to make a couple changes.
```

> First, your `html` variable will change to a function. The function will take in an argument, `todos`, which will populate the `window.todos` variable in the above code sample:

Reviewer: ```suggestion
First, your `html` variable will change to a function. The function will take in a `todos` argument, which will populate the `window.todos` variable in the above code sample:
```

`
```

> Your static page can take in `window.todos` and render HTML based on it, but you have not actually passed in any data from KV. To do this, you will need to make a couple changes.

Author: ```suggestion
Your static page can take in `window.todos` and render HTML based on it, but you have not actually passed in any data from KV. To do this, you will need to make a few changes.
```

`
```

> In `handleRequest`, use the retrieved KV data to call the `html` function, and generate a `Response` based on it:

Reviewer: ```suggestion
In `handleRequest`, use the retrieved KV data to call the `html` function and generate a `Response` based on it:
```

### Adding todos from the UI

> At this point, you have built a Cloudflare Worker that takes data from Cloudflare KV and renders a static page based on that Worker. That static page reads data, and generates a todo list based on that data. The remaining piece is creating todos, from inside the UI. You know that you can add todos using the KV API — you could update the cache by saying `TODOS.put(newData)`, and now you will learn how to update it from inside the UI.

Reviewer: ```suggestion
At this point, you have built a Cloudflare Worker that takes data from Cloudflare KV and renders a static page based on that Worker. That static page reads data and generates a todo list based on that data. The remaining piece is creating todos from inside the UI. You know that you can add todos using the KV API — you could update the cache by saying `TODOS.put(newData)`.
```

> To implement this, you will add a second handler in our Workers script, designed to watch for `PUT` requests to `/`. When a body is received at that URL, the worker will send the new todo data to our KV store.

Reviewer: ```suggestion
To update a todo item, you will add a second handler in our Workers script, designed to watch for `PUT` requests to `/`. When a request body is received at that URL, the worker will send the new todo data to our KV store.
```

}
```

> The script is pretty straightforward – you check that the request is a `PUT`, and wrap the remainder of the code in a `try/catch` block. First, you parse the body of the request coming in, ensuring that it is JSON, before you update the cache with the new data, and return it to the user. If anything goes wrong, simply return a 500. If the route is hit with an HTTP method other than `PUT` — that is, `POST`, `DELETE`, or anything else — return a 404.

Reviewer: ```suggestion
The script is pretty straightforward – you check that the request is a `PUT` and wrap the remainder of the code in a `try/catch` block. First, you parse the body of the request coming in, ensuring that it is JSON, before you update the cache with the new data and return it to the user. If anything goes wrong, simply return a `500` status code. If the route is hit with an HTTP method other than `PUT` — for example, `POST` or `DELETE` — return a `404` error.
```

> With this script, you can now add some “dynamic” functionality to your HTML page to actually hit this route. First, create an input for your todo “name”, and a button for “submitting” the todo.

Reviewer: ```suggestion
With this script, you can now add some “dynamic” functionality to your HTML page to actually hit this route. First, create an input for your todo “name” and a button for “submitting” the todo.
```

`
```

With the client-side code in place, deploying the new version of the function should put all these pieces together. The result is an actual dynamic todo list.

### Updating todos from the application UI

For the final piece of our todo list, you need to be able to update todos — specifically, marking them as completed.

> Luckily, a great deal of the infrastructure for this work is already in place. You can currently update the todo list data in our cache, as evidenced by our `createTodo` function. Performing updates on a todo, in fact, is much more of a client-side task than a Worker-side one.

Author: ```suggestion
Luckily, a great deal of the infrastructure for this work is already in place. You can update the todo list data in the cache, as evidenced by our `createTodo` function. Performing updates on a todo, in fact, is much more of a client-side task than a Worker-side one.
```

`
```

> So far, you have designed the client-side part of this code to take an array of todos in, and given that array, render out a list of simple HTML elements. There’s a number of things that you have been doing that you have not quite had a use for, yet: specifically, the inclusion of IDs, and updating the completed value on a todo. Luckily, these things work well together, in order to support actually updating todos in the UI.

Reviewer: ```suggestion
So far, you have designed the client-side part of this code to handle an array of todos and render a list of simple HTML elements. There’s a number of things that you have been doing that you have not quite had a use for yet – specifically, the inclusion of IDs and updating the todo's completed state. Luckily, these things work well together in order to support actually updating todos in the UI.
```

> To start, it would be useful to signify the ID of each todo in the HTML. By doing this, you can then refer to the element later, in order to correspond it to the todo in the JavaScript part of our code. Data attributes, and the corresponding `dataset` method in JavaScript, are a perfect way to implement this. When you generate your `div` element for each todo, you can attach a data attribute called todo to each `div`:

Reviewer: ```suggestion
To start, it would be useful to signify the ID of each todo in the HTML. By doing this, you can then refer to the element later in order to correspond it to the todo in the JavaScript part of our code. Data attributes and the corresponding `dataset` method in JavaScript are a perfect way to implement this. When you generate your `div` element for each todo, you can attach a data attribute called todo to each `div`:
```

`
```

So far, you have designed the client-side part of this code to handle an array of todos and render a list of simple HTML elements. There is a number of things that you have been doing that you have not quite had a use for yet – specifically, the inclusion of IDs and updating the todo's completed state. Luckily, these things work well together to actually support updating todos in the application UI.

> To start, it would be useful to signify the ID of each todo in the HTML. By doing this, you can then refer to the element later in order to correspond it to the todo in the JavaScript part of our code. Data attributes and the corresponding `dataset` method in JavaScript are a perfect way to implement this. When you generate your `div` element for each todo, you can attach a data attribute called todo to each `div`:

Author: ```suggestion
To start, it would be useful to attach the ID of each todo in the HTML. By doing this, you can then refer to the element later in order to correspond it to the todo in the JavaScript part of our code. Data attributes and the corresponding `dataset` method in JavaScript are a perfect way to implement this. When you generate your `div` element for each todo, you can attach a data attribute called todo to each `div`:
```

`
```

> Inside your HTML, each `div` for a todo now has an attached data attribute, which looks like:

Reviewer: I'll fix the code snippet, because it's outside of the PR changes itself.

Note to self:

```js
window.todos = []
var todoContainer = document.querySelector("#todos")
window.todos.forEach(todo => {
  var el = document.createElement("div")
  el.textContent = todo.name
  todoContainer.appendChild(el)
})
```

> Inside your HTML, each `div` for a todo now has an attached data attribute, which looks like:

Reviewer: Updated the code block to resolve this

`
```

> The checkbox is set up to correctly reflect the value of completed on each todo, but it does not yet update when you actually check the box. To do this, add an event listener on the `click` event, calling `completeTodo`. Inside the function, inspect the checkbox element, finding its parent (the todo div), and using the `todo` data attribute on it to find the corresponding todo in our data. Given that todo, you can toggle the value of completed, update our data, and re-render the UI:

Reviewer: ```suggestion
The checkbox is set up to correctly reflect the value of completed on each todo, but it does not yet update when you actually check the box. To do this, attach the `completeTodo` function as an event listener on the `click` event. Inside the function, inspect the checkbox element, find its parent (the todo div), and use its `todo` data attribute to find the corresponding todo in the data array. You can toggle the completed status, update its properties, and rerender the UI:
```

## Conclusions and next steps

> By completing this tutorial, you have created a pretty remarkable project: an almost entirely static HTML/JS application, transparently powered by Cloudflare KV and Workers, served on Cloudflare's edge network. There is a number of additions that can be made to this application, whether you want to implement a better design (you can refer to a live version at [todos.signalnerve.workers.dev](https://todos.signalnerve.workers.dev/)), or improvements to security, speed, etc.

Reviewer: transparently?

```suggestion
By completing this tutorial, you have created a pretty remarkable project. You have built a static HTML, CSS, and JavaScript application that is transparently powered by Workers and Workers KV and Workers, which takes full advantages of Cloudflare's edge network. 

There's room for improvement, too, if you feel so inclined. For example, you may want to implement a better design (you can refer to a live version at [todos.signalnerve.workers.dev](https://todos.signalnerve.workers.dev/)), or make additional improvements to security, speed, etc.
```

> Another addition is implementing per-user caching. Right now, the cache key is simply “data”: anyone visiting the site will share a todo list with any other user. Because you have the request information inside of your Worker, it is easy to make this data user-specific. For instance, implementing per-user caching by generating the cache key based on the requesting IP:

Reviewer: ```suggestion
You may also want to add user-specific caching. Right now, the cache key is always “data” – this means that any visitor to the site will share a same todo list with other visitors. Within your Worker, you could use values from the client request to create and maintain user-specific lists. For example, you may generate a cache key based on the requesting IP:
```

}
```

> One more deployment of your Workers project, and you have a full todo list application, with per-user functionality, served on Cloudflare's edge network.

Reviewer: ```suggestion
After making these changes and deploying the Worker one more time, your todo list application now includes per-user functionality while still taking full advantage of Cloudflare's edge network.
```
